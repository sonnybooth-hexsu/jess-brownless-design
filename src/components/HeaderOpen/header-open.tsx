import React, { useState } from "react"
import { Link } from "gatsby"
import { X, ChevronDown, ChevronUp } from "react-feather"
import logo from "../../assets/icons/logo.svg"
import styles from "./header-open.module.css"
import containerClasses from "../../styles/container.module.css"

type ProductItemsProps = {
  aestheticServices?: object
  oculoplasticServices?: object
  conditions?: object
}

type HeaderOpenProps = {
  siteTitle: string
  navLinks: Array<{
    name: string
    page: string
    id: number
    products?: boolean
    subLinks?: Array<{ name: string; page: string; id: number }>
  }>
  productItems: ProductItemsProps
  navToggle: () => Function
}

const SubMenuItem = ({ name, subLinks, id }) => {
  const [openState, setOpenState] = useState(false)

  return (
    <li
      key={`navigation-item-${id}`}
      data-testid="subLinkMenu"
      className={`animate-reveal uiRegular ${styles.headerNavigationLink} delay-${id}`}
      onClick={() => setOpenState(!openState)}
    >
      <span className={styles.headerNavigationSubMenu}>
        {name}
        {openState ? <ChevronUp /> : <ChevronDown />}
      </span>
      <div
        data-testid="subLinkMenuList"
        className={`${openState ? "block" : "hidden"}`}
      >
        {subLinks &&
          subLinks.map(({ name, page, id }) => (
            <a
              className="uiRegular"
              key={`navigation-sublink-item-${id}`}
              href={page}
            >
              {name}
            </a>
          ))}
      </div>
    </li>
  )
}

export const HeaderOpen = ({
  siteTitle,
  navLinks,
  navToggle,
  productItems,
}: HeaderOpenProps) => {
  return (
    <div className={`${styles.headerContainer} fadedBorder`}>
      <div className={`${styles.header} ${containerClasses.container}`}>
        <Link className={styles.headerLogo} to="/">
          <img className={styles.headerLogoImage} src={logo} alt={siteTitle} />
        </Link>
        <X className={styles.headerMenuIcon} onClick={navToggle} />
      </div>
      <nav
        className={`${styles.headerNavigation} ${containerClasses.container}`}
      >
        <ol>
          {navLinks.map(({ name, page, subLinks, id, products }) => {
            if (products) {
              let data

              // @ts-ignore: graphql data
              if (products === "conditions") {
                data = productItems.conditions
              }

              // @ts-ignore: graphql data
              if (products === "oculoplasticServices") {
                data = productItems.oculoplasticServices
              }

              // @ts-ignore: graphql data
              if (products === "aestheticServices") {
                data = productItems.aestheticServices
              }

              data.edges.sort((a, b) =>
                a.node.childMarkdownRemark.frontmatter.id >
                b.node.childMarkdownRemark.frontmatter.id
                  ? 1
                  : -1
              )

              const productsDestructured = data.edges.map(
                obj =>
                  // @ts-ignore: graphql data
                  obj.node.childMarkdownRemark.frontmatter
              )

              const productsDestructuredSubLinks = productsDestructured.map(
                ({ title, path, ...rest }) => {
                  return {
                    name: title,
                    page: path,
                    ...rest,
                  }
                }
              )

              return (
                <SubMenuItem
                  key={`navigation-item-${id}`}
                  name={name}
                  subLinks={productsDestructuredSubLinks}
                  id={id}
                />
              )
            }

            return subLinks && subLinks.length > 0 ? (
              <SubMenuItem
                key={`navigation-item-${id}`}
                name={name}
                subLinks={subLinks}
                id={id}
              />
            ) : (
              <li
                className={`animate-reveal ${styles.headerNavigationLink} delay-${id}`}
                data-testid="headerOpenNavigationLink"
                key={`navigation-item-${id}`}
              >
                <a className="uiRegular" href={`${page}`}>
                  {name}
                </a>
              </li>
            )
          })}
          <li
            className={`animate-reveal uiRegular delay-${navLinks.length + 1}`}
            key={`navigation-item-${navLinks.length + 1}`}
          >
            <a className="uiRegular" href="/contact">
              Contact
            </a>
          </li>
        </ol>
      </nav>
    </div>
  )
}
