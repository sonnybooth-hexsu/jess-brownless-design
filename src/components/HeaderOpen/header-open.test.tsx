/* eslint no-undef: 0 */

import React from "react"
import { render, cleanup } from "@testing-library/react"
import { queryAllByTestId, fireEvent } from "@testing-library/dom"
import { HeaderOpen } from "./header-open"

const testNavLinks = [
  {
    name: `Link 1`,
    page: `link-1`,
    subLinks: [
      {
        name: `Sub Link 1`,
        page: `Sub link-1`,
        id: 1,
      },
    ],
    id: 1,
  },
  {
    name: `Link 2`,
    page: `link-2`,
    subLinks: [
      {
        name: `Sub Link 1`,
        page: `Sub link-1`,
        id: 1,
      },
    ],
    id: 2,
  },
  {
    name: `Link 3`,
    page: `link-3`,
    subLinks: [],
    id: 3,
  },
]

const noneSubLinks = testNavLinks.filter(item => !(item.subLinks.length > 0))
const hasSubLinks = testNavLinks.filter(item => item.subLinks.length > 0)

afterEach(() => {
  cleanup()
})

describe("<HeaderOpen />", () => {
  it("should match snapshot", () => {
    const { container } = render(
      <HeaderOpen
        siteTitle="Test Site"
        navLinks={testNavLinks}
        navToggle={() => Function}
        productItems={{}}
      />
    )

    expect(container).toMatchSnapshot()
  })

  it("should have the length of links passed through", () => {
    const { container } = render(
      <HeaderOpen
        siteTitle="Test Site"
        navLinks={testNavLinks}
        navToggle={() => Function}
        productItems={{}}
      />
    )

    expect(
      queryAllByTestId(container, "headerOpenNavigationLink")
    ).toHaveLength(noneSubLinks.length)
  })

  it("toggles subLink class", () => {
    const { container } = render(
      <HeaderOpen
        siteTitle="Test Site"
        navLinks={testNavLinks}
        navToggle={() => Function}
        productItems={{}}
      />
    )

    const subLinks = queryAllByTestId(container, "subLinkMenu")
    expect(subLinks).toHaveLength(hasSubLinks.length)

    const sublinkMenuList = queryAllByTestId(subLinks[0], "subLinkMenuList")[0]

    expect(sublinkMenuList).toHaveClass("hidden")

    fireEvent.click(sublinkMenuList)

    expect(sublinkMenuList).toHaveClass("block")
  })
})
