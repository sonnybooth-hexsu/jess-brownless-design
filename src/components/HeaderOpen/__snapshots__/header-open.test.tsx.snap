// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HeaderOpen /> should match snapshot 1`] = `
<div>
  <div
    class="headerContainer fadedBorder"
  >
    <div
      class="header container"
    >
      <a
        class="headerLogo"
        href="/"
      >
        <img
          alt="Test Site"
          class="headerLogoImage"
          src="test-file-stub"
        />
      </a>
      <svg
        class="headerMenuIcon"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <line
          x1="18"
          x2="6"
          y1="6"
          y2="18"
        />
        <line
          x1="6"
          x2="18"
          y1="6"
          y2="18"
        />
      </svg>
    </div>
    <nav
      class="headerNavigation container"
    >
      <ol>
        <li
          class="animate-reveal uiRegular headerNavigationLink delay-1"
          data-testid="subLinkMenu"
        >
          <span
            class="headerNavigationSubMenu"
          >
            Link 1
            <svg
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <polyline
                points="6 9 12 15 18 9"
              />
            </svg>
          </span>
          <div
            class="hidden"
            data-testid="subLinkMenuList"
          >
            <a
              class="uiRegular"
              href="Sub link-1"
            >
              Sub Link 1
            </a>
          </div>
        </li>
        <li
          class="animate-reveal uiRegular headerNavigationLink delay-2"
          data-testid="subLinkMenu"
        >
          <span
            class="headerNavigationSubMenu"
          >
            Link 2
            <svg
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <polyline
                points="6 9 12 15 18 9"
              />
            </svg>
          </span>
          <div
            class="hidden"
            data-testid="subLinkMenuList"
          >
            <a
              class="uiRegular"
              href="Sub link-1"
            >
              Sub Link 1
            </a>
          </div>
        </li>
        <li
          class="animate-reveal headerNavigationLink delay-3"
          data-testid="headerOpenNavigationLink"
        >
          <a
            class="uiRegular"
            href="link-3"
          >
            Link 3
          </a>
        </li>
        <li
          class="animate-reveal uiRegular delay-4"
        >
          <a
            class="uiRegular"
            href="/contact"
          >
            Contact
          </a>
        </li>
      </ol>
    </nav>
  </div>
</div>
`;
